File: virtual_node_miner/tools/delete_add_edge.py
'''
    对数据集进行增删边操作，并保证不会将唯一一条出边删除
'''
import sys
import time
import random

# 增删边数据: 有的边增加或删除了两次
def deal_edge(file_path, chage_rate):
    update_path = file_path[:file_path.rindex('_base')]+'_update.e'  # 增删的边
    updated_path = file_path[:file_path.rindex('_base')]+'_updated.e' # 增删之后的边
    print('read file...', file_path)
    print('write file...', update_path)
    print('write file...', updated_path)
    f_w_update = open(update_path, 'w+', encoding='utf-8')
    f_w_updated = open(updated_path, 'w+', encoding='utf-8')
    f_r = open(file_path, 'r+', encoding='utf-8')
    have_out_edge = set() # 有出度的顶点
    all_num = 0
    add_num = 0
    del_num = 0
    while True:
        line = f_r.readline()
        if line == None or len(line) < 2:
            break
        all_num += 1
        num_list = line.strip('\t').strip(' ').strip('\n').split(' ')
        if(len(num_list) != 2):
            print("line check: ", len(num_list), num_list)
            exit(0)
        u = int(num_list[0])
        v = int(num_list[1])
        if u not in have_out_edge:
            have_out_edge.add(u) # 标记为有出边
            f_w_updated.write(line)
        else:
            p = random.random()
            if p < chage_rate:
                is_add = random.randint(0,1)
                if is_add == 0: # delete this edge
                    f_w_update.write('d ' + line) # 记录变化
                    del_num += 1
                else:
                    if len(have_out_edge) >= 2:
                        u, v = random.sample(have_out_edge, 2)
                        f_w_update.write('a ' + str(u) + ' ' + str(v) + '\n') # 记录变化
                        f_w_updated.write(str(u) + ' ' + str(v) + '\n') # 加入的边
                        add_num += 1
                    f_w_updated.write(line) # 此边不删除也要放入
            else:
                f_w_updated.write(line)
    print('all_num=%d add_num=%d del_num=%d update=%d rate=%f' % (all_num, add_num, del_num, add_num+del_num, (add_num+del_num)/all_num * 1.0))

    f_r.close()
    f_w_update.close()
    f_w_updated.close()

# 增删边数据
def deal_edge_2(file_path, chage_rate):
    update_path = file_path[:file_path.rindex('_base')]+'_update.e'  # 增删的边
    updated_path = file_path[:file_path.rindex('_base')]+'_updated.e' # 增删之后的边
    print('read file...', file_path)
    print('write file...', update_path)
    print('write file...', updated_path)
    f_w_update = open(update_path, 'w+', encoding='utf-8')
    f_w_updated = open(updated_path, 'w+', encoding='utf-8')
    f_r = open(file_path, 'r+', encoding='utf-8')
    have_out_edge = set() # 有出度的顶点
    all_num = 0
    add_num = 0
    del_num = 0
    origin_graph = {}
    # 读入原图
    while True:
        line = f_r.readline()
        if line == None or len(line) < 2:
            break
        all_num += 1
        num_list = line.strip('\t').strip(' ').strip('\n').split(' ')
        if(len(num_list) != 2):
            print("line check: ", len(num_list), num_list)
            exit(0)
        u = int(num_list[0])
        v = int(num_list[1])
        if u not in origin_graph:
            origin_graph[u] = set()
        origin_graph[u].add(v)
    
    # 进行增删
    should_delete_num = all_num * chage_rate
    nodes = list(origin_graph.keys())
    for k, adj in origin_graph.items():
        origin_graph[k] = list(adj)
    is_add = True
    while should_delete_num > 0:
        should_delete_num -= 1
        if is_add == False: # delete this edge
            cnt = 20
            is_add = True
            while cnt > 0:
                cnt -= 1
                x = random.choice(nodes)
                if len(origin_graph[x]) < 2:
                    continue
                y = random.choice(origin_graph[x])
                origin_graph[x].remove(y)
                f_w_update.write('d ' + str(x) + " " + str(y) + '\n') # 记录变化
                del_num += 1
                break
        else:
            cnt = 20
            is_add = False
            while cnt > 0:
                cnt -= 1
                u, v = random.choices(nodes, k=2)
                if v in origin_graph[u]:
                    continue
                f_w_update.write('a ' + str(u) + ' ' + str(v) + '\n') # 记录变化
                origin_graph[u].append(v)
                add_num += 1
    print('all_num=%d add_num=%d del_num=%d update=%d rate=%f' % (all_num, add_num, del_num, add_num+del_num, (add_num+del_num)/all_num * 1.0))
    for u, adj in origin_graph.items():
        for v in adj:
            f_w_updated.write(str(u) + ' ' + str(v) + '\n') 
    f_r.close()
    f_w_update.close()
    f_w_updated.close()

# 测试增删是否正确
def check(file_path):
    update_path = file_path[:file_path.rindex('_base')]+'_update.e' # 增删之后的边
    updated_path = file_path[:file_path.rindex('_base')]+'_updated.e' # 增删之后的边
    print('read file...', file_path)
    print('read file...', updated_path)
    f_r_update = open(update_path, 'r+', encoding='utf-8')
    f_r_updated = open(updated_path, 'r+', encoding='utf-8')
    f_r = open(file_path, 'r+', encoding='utf-8')
    origin_graph = {}
    update_graph = {}
    while True:
        line = f_r.readline()
        if line == None or len(line) < 2:
            break
        num_list = line.strip('\t').strip(' ').strip('\n').split(' ')
        if(len(num_list) != 2):
            print("line check: ", len(num_list), num_list)
            exit(0)
        u = int(num_list[0])
        v = int(num_list[1])
        if u not in origin_graph:
            origin_graph[u] = set()
        origin_graph[u].add(v)
    # 根新新图
    while True:
        line = f_r_update.readline()
        if line == None or len(line) < 2:
            break
        num_list = line.strip('\t').strip(' ').strip('\n').split(' ')
        if(len(num_list) != 3):
            print("line check: ", len(num_list), num_list)
            exit(0)
        type = num_list[0]
        u = int(num_list[1])
        v = int(num_list[2])
        if u not in origin_graph:
            print("get new graph error.", line)
            exit(0)
        if type == 'a':
            origin_graph[u].add(v)
        elif type == 'd':
            origin_graph[u].remove(v)
        else:
            print('type error', line)
    # 直接读新图
    while True:
        line = f_r_updated.readline()
        if line == None or len(line) < 2:
            break
        num_list = line.strip('\t').strip(' ').strip('\n').split(' ')
        if(len(num_list) != 2):
            print("line check: ", len(num_list), num_list)
            exit(0)
        u = int(num_list[0])
        v = int(num_list[1])
        if u not in update_graph:
            update_graph[u] = set()
        update_graph[u].add(v)

    if len(list(origin_graph.keys())) != len(list(update_graph.keys())):
        print('num not same.')
        exit(0)
    for k, adj in origin_graph.items():
        # for u in adj:
        if adj != update_graph[k]:
            print('not same.', k, adj, update_graph[k])
            print('diff set:', adj - update_graph[k])
            print('diff set:', update_graph[k] - adj)
            exit(0)
    print('same....')
    f_r.close()
    f_r_updated.close()
    f_r_update.close()

if __name__ == "__main__":
    # 运行命令： python3 ./tools/delete_add_edge.py /home/yusong/dataset/web-uk-2005/web-uk-2005_base.e 0.01
    # 读入文件路径
    print('参数列表:', str(sys.argv))
    # file_path = input().strip(' ')

    file_path = sys.argv[1].strip(' ')
    chage_rate = float(sys.argv[2].strip(' '))
    print('start deal data...')
    
    time_start=time.time()
    # deal_edge_2(file_path, chage_rate)   #  处理数据
    time_end=time.time()
    print('delete_add_edge finish!')  
    print('time cost',time_end-time_start,'s')

    # 检测是否正确
    time_start=time.time()
    check(file_path)
    time_end=time.time()
    print('check time cost',time_end-time_start,'s')

File: virtual_node_miner/tools/delete_add_edge.cc
/*
 * 进行边的增删
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <unordered_set>
#include <unordered_map>
#include <ctime>
#include <string>
#include <stdlib.h>
#include <cmath>
#include <cassert>
#include <algorithm>
#include "../utils/timer.h"

using adjlist = std::vector<int>;

class Delete_add_edge
{
public:
    std::vector<adjlist> _adjlists;
    size_t _num_node;
    size_t _raw_num_node;
    size_t _num_edge;
    size_t _raw_num_edge;
    std::unordered_map<int, int> vertex_map; //原id: 新id
    std::unordered_map<int, int> vertex_reverse_map; // 新id: 原id

    bool load_graph(const std::string &file_path) {
        auto load_start = time(nullptr);

        FILE *f = fopen(file_path.c_str(), "r");

        if (f == 0) {
            std::cout << "file cannot open! " << file_path << std::endl;
            return false;
        }

        std::vector<std::pair<int, int>> edges;

        int u = 0;
        int v = 0;
        _num_node = 0;
        while (fscanf(f, "%d %d", &u, &v) > 0) {
            assert(u >= 0);
            assert(v >= 0);
            // _num_node = std::max(_num_node, size_t(u + 1));
            // _num_node = std::max(_num_node, size_t(v + 1));
            // edges.emplace_back(std::pair<int, int>(u, v));
            if(vertex_map.find(u) == vertex_map.end()){
                vertex_map[u] = _num_node;
                vertex_reverse_map[_num_node] = u;
                u = _num_node++;
            }
            else{
                u = vertex_map[u];
            }
            if(vertex_map.find(v) == vertex_map.end()){
                vertex_map[v] = _num_node;
                vertex_reverse_map[_num_node] = v;
                v = _num_node++;
            }
            else{
                v = vertex_map[v];
            }
            edges.emplace_back(std::pair<int, int>(u, v));
        }
        fclose(f);

        std::sort(edges.begin(), edges.end());

        _num_edge = edges.size();

        _adjlists.clear();
        _adjlists.resize(_num_node);
        for (auto edge : edges) {
            _adjlists[edge.first].emplace_back(edge.second);
        }

        return true;
    }

    void run(const std::string &edge_path, const float rate){
        // web-uk-2002-all_base.e
        int pos = edge_path.find_last_of("b"); // 找到第一个空格的位置
        std::string base_path = edge_path.substr(0, pos);
        std::string updated_path = base_path + "updated.e";
        std::string update_path = base_path + "update.e";
        FILE *f_updated = fopen(updated_path.c_str(), "w");
        FILE *f_update = fopen(update_path.c_str(), "w");
        if (f_updated == nullptr) {
            std::cout << "graph file cannot create!" << updated_path << std::endl;
            exit(0);
        }
        if (f_update == nullptr) {
            std::cout << "graph file cannot create!" << update_path << std::endl;
            exit(0);
        }
        timer_next("load_data");
        load_graph(edge_path);
        int update_num = _num_edge * rate;
        bool is_add = true;

        printf("load finied, node_num=%d edge_num=%d\n", _num_node, _num_edge);

        std::vector<int>::iterator it;
        int u, v;
        int add_num = 0;
        int del_num = 0;
        std::cout << "update_num=" << update_num << std::endl;
        timer_next("add_and_del");
        while(update_num > 0){
            update_num--;
            if(is_add){ // add edge
                is_add = false;
                int cnt = 20;
                while(cnt--){
                    u = rand() % _num_node;
                    v = rand() % _num_node;
                    it = find(_adjlists[u].begin(), _adjlists[u].end(), v);
                    if(it == _adjlists[u].end()){
                        // printf("a %d %d cnt=%d\n", u, v, update_num);
                        // printf("a %d %d cnt=%d\n", vertex_reverse_map[u], vertex_reverse_map[v], update_num);
                        fprintf(f_update, "a %d %d\n", vertex_reverse_map[u], vertex_reverse_map[v]);
                        _adjlists[u].emplace_back(v);
                        add_num++;
                        break;
                    }
                }
            }
            else{ // delete edge
                is_add = true;
                int cnt = 20;
                while(cnt--){
                    u = rand() % _num_node;
                    if(_adjlists[u].size() > 2){
                        v = rand() % (_adjlists[u].size());
                        // printf("d %d %d\n", vertex_reverse_map[u], vertex_reverse_map[_adjlists[u][v]]);
                        fprintf(f_update, "d %d %d\n", vertex_reverse_map[u], vertex_reverse_map[_adjlists[u][v]]);
                        _adjlists[u][v] = _adjlists[u][_adjlists[u].size()-1];
                        _adjlists[u].pop_back();
                        del_num++;
                        break;
                    }
                }
            }
        }
        // 写入updated文件：
        timer_next("write_data");
        for (int u = 0; u < _num_node; u++) {
            for (int v : _adjlists[u]) {
                fprintf(f_updated, "%d %d\n", vertex_reverse_map[u], vertex_reverse_map[v]);
            }
        }
        fclose(f_updated);
        fclose(f_update);
        printf("all_num=%d add_num=%d del_num=%d rate=%f\n", _num_edge, add_num, del_num, (add_num+del_num)*1.0/_num_edge);
        std::cout << update_path << std::endl;
        std::cout << updated_path << std::endl;
    }
};

int main(int argc, char const *argv[])
{
    // ./a.out /home/yusong/dataset/web-uk-2002-all/web-uk-2002-all_base.e 0.01
    timer_start(true);
    std::string edge_path = argv[1];
    float rate = atof(argv[2]);

    Delete_add_edge deal = Delete_add_edge();
    deal.run(edge_path, rate);

    timer_end(false, "-del_add_edge");

    return 0;
}

